rfcs/192-fr-copy-receipt-performance.md

```diff
@@ -0,0 +1,186 @@
+}
+```
+
+## Migration
+
+The new table will be created with the usual migration process.
+
+On start up of the Queue (in the 1.2 `workerFR.cs`) we check if the table contains any entries and if not we will populate it with the old data from the `ftJournalFR` table.
+
+The population of the table will be done like this:
+
+```cs
+private async Task<int> PopulateFtJournalFRCopyReceiptsTableAsync()
+{
+    await foreach (var copyJournal in _journalFRRepository.GetProcessedCopyReceiptsAsync())
+    {
+        var queueItem = await _queueItemRepository.GetAsync(copyJournal.ftQueueItemId);
+        var request = queueItem?.request != null ? JsonConvert.DeserializeObject<ReceiptRequest>(queueItem.request) : null;
+        var response = queueItem?.response != null ? JsonConvert.DeserializeObject<ReceiptResponse>(queueItem.response) : null;
+        if (response != null)
+        {
+            var duplicata = response.ftSignatures.FirstOrDefault(x => x.Caption == "Duplicata");
+
+            if (duplicata != null)
+            {
+                _journalFRCopyReceiptsRepository.Insert(new FtJournalFRCopyReceipt
+                {
+                    cbPreviousReceiptReference = request.cbPreviousReceiptReference,
+                    ftQueueItemId = copyJournal.ftQueueItemId
+                });
+            }
+        }
+    }
+
+    return count;
+}
+```
+
+## SignatureCloud
+
+We will setup a new SignatureCloud environment which uses the new implementation and keep using the old one in the beginning.
+We will have a load balancer in front of the SignatureCloud which will route the requests to the correct environment.
+
+When the PosCreator is ready to use the new implementation we will switch the load balancer to the new environment.
+This will be done in coordination with the PosCreators and the recertification.
+
+### AzureTableStorage
+
+The Azure Table Storage Queue will have the `cbPreviousReceiptReference` as a Partition key and the `ftQueueItemID` as a Row key.
+
+This means counting all rows with the same `cbPreviousReceiptReference` will be very fast.
+
+# Drawbacks
+
+This solution involves a new table and we need to populate it with the data from the `ftJournalFR` table before we can use the new implementation.
+This brings some problems with it especially in the SignatureCloud.
+
+This update version of the middleware might need to be certified again.
+
+# Rationale and alternatives
+
+An alternative way of doing this would be to keep the current implementation and improve it.
+
+One way to do that would be to not count every CopyReceipt in the `ftJournalFR` table but only find the latest one and increase its count by one.
+
+If we find a non copy receipt with the same `cbReceiptReference` as the `cbPreviousReceiptReference` we can return 0 directly.
+
+```cs
+private async Task<int> GetCountOfExistingCopiesAsync(string cbPreviousReceiptReference)
```

Comment:
So IMO we should be good if we start including this fix for a first `rc`. Based on how this fix performs we can then decide how we build up the table ( maybe in process, or at the beginning).